contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+    schedule.exists = true;
+    schedule.startTime = scheduleData.startTime;
+    schedule.endTime = scheduleData.startTime + restrictionDuration.duration;
+    schedule.releasedAmountFloor = 0;
+    emit ScheduleCreated(projectId, schedule.startTime, schedule.endTime);
+  }
+
+  /**
+   * @notice Revokes unreleased tokens from a schedule.
+   * This is an *admin* operation callable only by addresses having TOKEN_REVOKER_ROLE
+   * (enforced in `batchRevokeUnreleasedTokenAmounts`)
+   *
+   * @dev Only the maximum revocable number of tokens (unreleased tokens) can be revoked.
+   * Once the tokens have been revoked, the current released amount can never fall below
+   * its current level, even if the linear release schedule of the new amount would cause
+   * the released amount to be lowered at the current timestamp (a floor is established).
+   *
+   * Unlike in the `withdrawFromSchedule` function, here we burn `RestrictedNORI`
+   * from the schedule owner but send that `BridgedPolygonNORI` back to Nori's
+   * treasury or an address of Nori's choosing (the *to* address).
+   * The *claimedAmount* is not changed because this is not a claim operation.
+   */
+  function _revokeUnreleasedTokens(
+    address to,
+    uint256 removalId,
+    uint256 amount
+  ) internal {
+    // slither-disable-next-line calls-loop
+    uint256 scheduleId = _removal.getProjectIdForRemoval(removalId);
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    if (!schedule.exists) {
+      revert NonexistentSchedule({scheduleId: scheduleId});
+    }
+    uint256 quantityToRevoke;
+    // amount of zero indicates revocation of all remaining tokens.
+    if (amount > 0) {
+      if (!(amount <= revocableQuantityForSchedule(scheduleId))) {
+        revert InsufficientUnreleasedTokens({scheduleId: scheduleId});
+      }
+      quantityToRevoke = amount;
+    } else {
+      quantityToRevoke = revocableQuantityForSchedule(scheduleId);
+    }
+
+    // burn correct proportion from each token holder
+    EnumerableSetUpgradeable.AddressSet storage tokenHolders = schedule
+      .tokenHolders;
+    uint256[] memory quantitiesToBurnForHolders = new uint256[](
+      tokenHolders.length()
+    );
+
+    uint256 cumulativeQuantityToBurn = 0;
+    for (uint256 i = 0; i < (tokenHolders.length() - 1); i++) {
+      uint256 quantityToBurnForHolder = _quantityToRevokePerTokenHolder(
+        quantityToRevoke,
+        scheduleId,
+        tokenHolders.at(i)
+      );
+      quantitiesToBurnForHolders[i] = quantityToBurnForHolder;
+      cumulativeQuantityToBurn += quantityToBurnForHolder;
+    }
+    quantitiesToBurnForHolders[tokenHolders.length() - 1] =
+      quantityToRevoke -
+      cumulativeQuantityToBurn;
+
+    for (uint256 i = 0; i < (tokenHolders.length()); i++) {
+      super._burn(
+        tokenHolders.at(i),
+        scheduleId,
+        quantitiesToBurnForHolders[i]
+      );
+      schedule.quantitiesRevokedByAddress[
```

Comment:
can this be moved to the previous for loop in addition to using burnBatch in the previous statement? Seems like this doesnt need to loop through token holders twice

Reply:
so I moved both `_burn` (since we can't use `burnBatch` here) and the state update into the one loop, also realizing that we don't need to store the quantities to burn for each token holder if we just burn it on the fly.  saved some gas according to reporter

Reply:
never mind... we can't do this because the math that figures out the correct quantity to burn for each token holder relies on the balances being unchanged while that math is being performed, so we really do need to calculate all the quantities before going on to burn them

Reply:
gotcha-- I think my comment [here](https://github.com/nori-dot-eco/contracts/pull/249/files/998a7c2f78de2bf6a1f41f2c5161c04e02449ab5..fb97bb8a727a24cdc034f908b27899c6a7b61e26#r906864378) might be able to provide a middle ground in the future by creating another batchBurn variant?

Reply:
added a todo